Don't panic on second call if server is off

chapa opened the pull request:
Transform `oneshot::Canceled` errors to `OperationError::ConnectionHasDropped` and forward them to the caller so it can handle them properly.

Previous `.unwrap()` used to cause panicking when two operations was requested on a shutdown server.

Fix #52 

### Additional information

I don't know why, but I wasn't able to test writing operations. They were all blocking the test that couldn't finish, not even a panic :thinking: 
- `connection.write_events`
- `connection.start_transaction`
- `transaction.write`
- `transaction.commit`
- `connection.delete_stream`
- `connection.update_persistent_subscription`
- `connection.delete_persistent_subscription`

Also, the test `offline_server_test` I wrote displays this "error" :
```
ERROR eventstore::internal::driver] Maximum reconnection attempt count reached (0)!
```
I don't know if it's disturbing ?

Another detail, we can't change the `Driver::reconnect_delay` with the `ConnectionBuilder` API, which makes the test 3 seconds too long... :yum:

chapa committed:
Improve error handling

Transform `oneshot::Canceled` errors to `OperationError::ConnectionHasDropped` and forward them to the caller so it can handle them properly.

Previous `.unwrap()` used to cause panicking when two operations was requested on a shutdown server.

YoEight commented:
Hi @chapa!
I just saw that PR, it didn't show up in my notification feed (I was about to fix that specific issue 😄).

YoEight commented on src/internal/operations.rs:
```diff
@@ -179,7 +179,7 @@ impl WriteEvents {
         });
 
-        promise.await.unwrap()
+        promise.await.map_err(|_| types::OperationError::ConnectionHasDropped)?
```
`ConnectionHasDropped` has a different meaning internally, In this situation, a better error would be `ConnectionClosed`. That enum variant doesn't exist yet and I suggest to add it.

YoEight commented on src/internal/operations.rs:
```diff
@@ -289,7 +289,7 @@ impl ReadEvent {
         });
 
-        promise.await.unwrap()
+        promise.await.map_err(|_| types::OperationError::ConnectionHasDropped)?
```
Same as above.

YoEight commented on src/internal/operations.rs:
```diff
@@ -409,7 +409,7 @@ impl TransactionStart {
         });
 
-        promise.await.unwrap()
+        promise.await.map_err(|_| types::OperationError::ConnectionHasDropped)?
```
Same as above.

YoEight commented on src/internal/operations.rs:
```diff
@@ -663,7 +663,7 @@ impl TransactionCommit {
         });
 
-        promise.await.unwrap()
+        promise.await.map_err(|_| types::OperationError::ConnectionHasDropped)?
```
Same as above.

YoEight commented on src/internal/operations.rs:
```diff
@@ -811,7 +811,7 @@ impl ReadStreamEvents {
         });
 
-        promise.await.unwrap()
+        promise.await.map_err(|_| types::OperationError::ConnectionHasDropped)?
```
Same as above.

YoEight commented on src/internal/operations.rs:
```diff
@@ -936,7 +936,7 @@ impl ReadAllEvents {
         });
 
-        promise.await.unwrap()
+        promise.await.map_err(|_| types::OperationError::ConnectionHasDropped)?
```
Same as above.

YoEight commented on src/internal/operations.rs:
```diff
@@ -1063,7 +1063,7 @@ impl DeleteStream {
         });
 
-        promise.await.unwrap()
+        promise.await.map_err(|_| types::OperationError::ConnectionHasDropped)?
```
Same as above.

YoEight commented on src/internal/operations.rs:
```diff
@@ -1864,7 +1864,7 @@ impl CreatePersistentSubscription {
         });
 
-        promise.await.unwrap()
+        promise.await.map_err(|_| types::OperationError::ConnectionHasDropped)?
```
Same as above.

YoEight commented on src/internal/operations.rs:
```diff
@@ -2053,7 +2053,7 @@ impl DeletePersistentSubscription {
         });
 
-        promise.await.unwrap()
+        promise.await.map_err(|_| types::OperationError::ConnectionHasDropped)?
```
Same as above.

YoEight commented on tests/integration.rs:
```diff
@@ -647,6 +647,33 @@ pub mod tcp {
+    fn offline_server_test() {
+        block_on(async {
+            let host = std::env::var("EVENTSTORE_HOST").unwrap_or("127.0.0.1".to_string());
+            let conn_str = format!("{}:1114", host); // Be sure your server doesn't run on port 1114
```
The server runs on 1113.

YoEight commented on tests/integration.rs:
```diff
@@ -647,6 +647,33 @@ pub mod tcp {
     }
+
+    #[test]
+    fn offline_server_test() {
```
I don't see how `offlive_server_test` captures the issue. The problem happens only when the server is offline and the client gave up trying to reconnect. I don't think a simple unit test will work here. We need to move to https://github.com/testcontainers/testcontainers-rs.

I don't ask you doing this in this PR (but feel free to do it if you want 😃 )

YoEight replied:
I think I understand what you did. You purposely choose a wrong port to trigger the issue at the first attempt. Forget what I said earlier then, it's a good idea!

YoEight commented on src/internal/operations.rs:
```diff
@@ -1972,7 +1972,7 @@ impl UpdatePersistentSubscription {
         });
 
-        promise.await.unwrap()
+        promise.await.map_err(|_| types::OperationError::ConnectionHasDropped)?
```
Same as above.

YoEight commented:
Thank you for your PR. It looks good minor a few remarks. ~I let a comment about the test. If you don't want to do it, please remove that test. A proper test would need a better testing infrastructure to capture that particular situation.~

Could you change the PR title, so it reflects the issue you opened earlier #52, please?

Somehow, no GitHub Actions were triggered. Let's hope your next commit push will trigger them.

YoEight commented:
I made `reconnect_delay` configurable now https://github.com/YoEight/eventstore-rs/pull/55, if you want to use it.

chapa committed:
Speed up offline_server_test

Set connection_timeout to 0 to stop waiting the connection to a known offline server

chapa committed:
Use ConnectionClosed rather than ConnectionHasDropped

chapa commented:
Hi, thanks for your feedback, I changed the PR title and did the modifications you suggests :wink:

chapa commented:
Strangely though, Github Actions doesn't seem to have been triggered this time either :thinking:

YoEight commented:
Hi @chapa,

I fixed the GitHub Action issue however you now have to rebase your branch from the master branch so your pull request can enjoy it. I apologise for the inconvenience.

chapa committed:
Fix linting warnings

YoEight commented:
Thank you for your work @chapa !

chapa commented:
You're welcome :wink:
